flashlight/fl/examples/RnnClassification.cpp

```diff
@@ -0,0 +1,260 @@
+
+// return a random int between [mini, maxi]
+int randi(int mini, int maxi) {
+  if(maxi<mini) 
+    std::swap(maxi,mini);
+  return rand()%(maxi-mini + 1) + mini; 
+}
+
+class ClassDataset : public Dataset {
+ public:
+  using names=std::vector<std::string>;
+  std::map<std::string, names> datasets;
+  static std::map<std::string, unsigned> label2ID;
+  static std::map<unsigned, std::string> ID2label;
+  unsigned totalnumexamples=0;
+
+  // read the folder/lang.txt file and register the examples in the datasets map
+  void read(const std::string folder, const std::string lang) {
+    FL_VLOG(fl::INFO) << "Opening " << folder << "/" << lang << ".txt ...";
+    std::ifstream file(folder + "/" + lang +".txt");
+    if (!file.is_open()) {
+      throw af::exception("Can't open the input dataset file");
+    }
+    unsigned id = label2ID.size();
+    label2ID[lang] = id;
+    ID2label[id] = lang;
+    names v;
+    std::string line;
+    while (std::getline(file, line)) {
+      if(line.empty()) continue;
+      v.push_back(line);
+    }
+    totalnumexamples += v.size();
+    std::cout << "Found " << v.size() << " examples for category " << lang << ". Total: " << totalnumexamples <<std::endl; 	
+    datasets[lang]=v;
+  }
+
+  // Turn a string into an AF array <1 x line_length> of char indices
+  static af::array lineToTensor(const std::string& line) {
+    std::vector<float> d;
+    for(char c : line) {
+      d.push_back((float)(c)); // direct cast of char to float
+    }
+    return af::array(1, d.size(), d.data());
+  }
+
+  ClassDataset(const std::string dataset_path) {
+    // Todo: search for all dataset files using std filesystem if cpp17 enabled
+    std::vector<std::string> lang = { "Arabic", "Greek", "Chinese", "Czech", "Dutch", "Japanese", "Korean", "Russian", "English", "Scottish", "Vietnamese", "German", "Spanish", "French", "Polish", "Italian", "Irish" };
+    for (auto& l : lang)
+      read(dataset_path, l);
+    for(auto& it : ID2label)
+      std::cout << it.first << ":" << it.second << ", ";
+    std::cout << std::endl;
+  }
+
+  // each epoch to go over some percent of the training dataset
+  int64_t size() const override { return .3f * totalnumexamples; }
+
+  // get a random example: name, category
+  std::pair<std::string, std::string> getRandomExample() const {
+    std::pair<std::string, std::string> p;
+    auto it=datasets.begin();
+    unsigned cat = randi(0, datasets.size()-1); // random category index
+    std::advance(it, cat);
+    p.second = it->first; // category name
+    const auto& v=it->second;
+    unsigned nn = v.size();
+    unsigned ri = randi(0,nn-1);
+    p.first = v[ri]; // name
+    FL_VLOG(fl::INFO) << "Random example:" << p.first << " class:" << cat << " : " << it->first;
+    return p;
+  }
+
+  // get a (random) example and return a vector of 2 tensors : the input and the expected category index
+  std::vector<af::array> get(const int64_t ) const override {
+    auto p = getRandomExample();
```

Comment:
does pytorch do with replacement?

Reply:
hum, which kind of replacement are you speaking about ?

Reply:
random sample picking during epoch, seems right now it is with replacement, so the same sample can appear several times during the same epoch

Reply:
I see: even if pytorch seems to have a  'WeightedRandomSampler' with a priori some support for replacement, the linked pytorch tutorial does not do with replacement.